### PR TITLE
lint: sort imports

### PR DESCRIPTION
**Pull request by daixiang0**
Signed-off-by: Loong Dai <loong.dai@intel.com>

Thank you for contributing to Harbor!

# Comprehensive Summary of your change

Add goimprots linter to group Go imports.

Please indicate you've done the following:
- [x] Well Written Title and Summary of the PR
- [x] Label the PR as needed. "release-note/ignore-for-release, release-note/new-feature, release-note/update, release-note/enhancement, release-note/community, release-note/breaking-change, release-note/docs, release-note/infra, release-note/deprecation"
- [x] Accepted the DCO. Commits without the DCO will delay acceptance.
- [ ] Made sure tests are passing and test coverage is added if needed.
- [ ] Considered the docs impact and opened a new docs issue or PR with docs changes if needed in [website repository](https://github.com/goharbor/website).

### CODE REVIEW

**Comment by daixiang0** [comment]
release-note/ignore-for-release

**Comment by zyyw** [comment]
@daixiang0 Thanks you for the contribution to Harbor project.
Could you please update the import into this format, separating the import into 3 groups:
* import built-in packages
* import the current goharbor/harbor project packages, i.e. "github.com/goharbor/harbor"
* import third-party packages
and sort the import within each group
for example:
```
import (
	"net/http"
	"os"
	"strings"

	"github.com/goharbor/harbor/src/common/dao"
	commonthttp "github.com/goharbor/harbor/src/common/http"
	"github.com/goharbor/harbor/src/common/models"
	"github.com/goharbor/harbor/src/lib/log"
	"github.com/goharbor/harbor/src/pkg/exporter"

	_ "github.com/jackc/pgx/v4/stdlib" // registry pgx driver
	"github.com/prometheus/client_golang/prometheus"
	"github.com/spf13/viper"
)
```

**Comment by daixiang0** [comment]
Sure, the GCI would this work.

**Comment by zyyw** [comment]
Hi @daixiang0 , do we have any update on this PR.

**Comment by daixiang0** [comment]
Will update later, thanks.

**Comment by daixiang0** [comment]
Now linter only supports:

- import built-in packages
- import third-party packages
- import the current goharbor/harbor project packages, i.e. "github.com/goharbor/harbor"

I will update if accepted.

**Comment by zyyw** [comment]
> Now linter only supports:
> 
> * import built-in packages
> * import third-party packages
> * import the current goharbor/harbor project packages, i.e. "github.com/goharbor/harbor"
> 
> @zyyw I will update if ok.
It will also be grouped into 3 parts, right?
If so, that's also fine with me. Please update it. Thank you for the contribution!

**Comment by daixiang0** [comment]
Sure.

File: src/pkg/oidc/testutils.go [hunk 1]
```diff
@@ -2,11 +2,9 @@ package oidc
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"strconv"
-)
```
Comment (zyyw) [inline src/pkg/oidc/testutils.go:7]: This close parentheses should NOT be removed.

File: src/chartserver/cache.go [hunk 2]
```diff
@@ -7,10 +7,10 @@ import (
 	"time"
 
 	beego_cache "github.com/beego/beego/cache"
-	hlog "github.com/goharbor/harbor/src/lib/log"
-
 	// Enable redis cache adaptor
 	_ "github.com/beego/beego/cache/redis"
+
```
Comment (zyyw) [inline src/chartserver/cache.go:12]: Do we need this empty line 12 here? Maybe we should remove it.

File: src/chartserver/handler_repo_test.go [hunk 3]
```diff
@@ -1,7 +1,6 @@
 package chartserver
 
 import "testing"
-
```
Comment (zyyw) [inline src/chartserver/handler_repo_test.go:4]: Why removing this line 4 here? Is it deleted by `make lint`?  Maybe we should keep it.

File: src/pkg/oidc/testutils.go (hunk 1 above)
Comment (daixiang0) [reply src/pkg/oidc/testutils.go:7]: Seems linter's bug, let me fix it.

File: src/chartserver/handler_repo_test.go (hunk 3 above)
Comment (zyyw) [reply src/chartserver/handler_repo_test.go:4]: Essentially, we would like there is an empty line between `import` and `func TestGetIndexFile`. If this  line 4 is deleted automatically by `make lint`,  maybe we can try to change 
```
import "testing"

// Test get /index.yaml
func TestGetIndexFile(t *testing.T) {
...
}
```
to 
```
import (
    "testing"
)

// Test get /index.yaml
func TestGetIndexFile(t *testing.T) {
...
}
```
to  ensure there is an empty line between `import` and its subsequent code.

Comment (daixiang0) [reply src/chartserver/handler_repo_test.go:4]: Got it, I will try to update.

File: src/controller/event/handler/replication/event/event.go [hunk 4]
```diff
@@ -15,7 +15,6 @@
 package event
 
 import "github.com/goharbor/harbor/src/pkg/reg/model"
-
```
Comment (zyyw) [inline src/controller/event/handler/replication/event/event.go:18]: similar comment to this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/controller/event/model/event.go [hunk 5]
```diff
@@ -1,7 +1,6 @@
 package model
 
 import "github.com/goharbor/harbor/src/pkg/retention/policy/rule"
-
```
Comment (zyyw) [inline src/controller/event/model/event.go:4]: same: 
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/controller/jobservice/model.go [hunk 6]
```diff
@@ -15,7 +15,6 @@
 package jobservice
 
 import "time"
-
```
Comment (zyyw) [inline src/controller/jobservice/model.go:18]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/controller/replication/flow/mock.go [hunk 7]
```diff
@@ -15,7 +15,6 @@
 package flow
 
 import "github.com/goharbor/harbor/src/pkg/reg/adapter"
-
```
Comment (zyyw) [inline src/controller/replication/flow/mock.go:18]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/core/service/notifications/base.go [hunk 8]
```diff
@@ -1,7 +1,6 @@
 package notifications
 
 import "github.com/goharbor/harbor/src/core/api"
-
```
Comment (zyyw) [inline src/core/service/notifications/base.go:4]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/jobservice/job/impl/legacy/legacy.go [hunk 9]
```diff
@@ -15,7 +15,6 @@
 package legacy
 
 import "github.com/goharbor/harbor/src/pkg/scheduler"
-
```
Comment (zyyw) [inline src/jobservice/job/impl/legacy/legacy.go:18]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/jobservice/logger/backend/std_logger_test.go [hunk 10]
```diff
@@ -1,7 +1,6 @@
 package backend
 
 import "testing"
-
```
Comment (zyyw) [inline src/jobservice/logger/backend/std_logger_test.go:4]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/jobservice/period/mock_scheduler.go [hunk 11]
```diff
@@ -3,7 +3,6 @@
 package period
 
 import mock "github.com/stretchr/testify/mock"
-
```
Comment (zyyw) [inline src/jobservice/period/mock_scheduler.go:6]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/lib/authorizer.go [hunk 12]
```diff
@@ -15,6 +15,5 @@
 package lib
 
 import "github.com/goharbor/harbor/src/common/http/modifier"
-
```
Comment (zyyw) [inline src/lib/authorizer.go:18]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/lib/config/metadata/metadatalist.go [hunk 13]
```diff
@@ -15,7 +15,6 @@
 package metadata
 
 import "github.com/goharbor/harbor/src/common"
-
```
Comment (zyyw) [inline src/lib/config/metadata/metadatalist.go:18]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/lib/encode/repository/repository.go [hunk 14]
```diff
@@ -1,7 +1,6 @@
 package repository
 
 import "net/url"
-
```
Comment (zyyw) [inline src/lib/encode/repository/repository.go:4]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/lib/orm/creator.go [hunk 15]
```diff
@@ -15,7 +15,6 @@
 package orm
 
 import "github.com/beego/beego/orm"
-
```
Comment (zyyw) [inline src/lib/orm/creator.go:18]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/lib/response_recorder.go [hunk 16]
```diff
@@ -15,7 +15,6 @@
 package lib
 
 import "net/http"
-
```
Comment (zyyw) [inline src/lib/response_recorder.go:18]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/pkg/notifier/notification_handler.go [hunk 17]
```diff
@@ -1,7 +1,6 @@
 package notifier
 
 import "context"
-
```
Comment (zyyw) [inline src/pkg/notifier/notification_handler.go:4]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/pkg/quota/types/format_test.go [hunk 18]
```diff
@@ -15,7 +15,6 @@
 package types
 
 import "testing"
-
```
Comment (zyyw) [inline src/pkg/quota/types/format_test.go:18]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/pkg/reg/adapter/aliacr/types.go [hunk 19]
```diff
@@ -1,7 +1,6 @@
 package aliacr
 
 import "time"
-
```
Comment (zyyw) [inline src/pkg/reg/adapter/aliacr/types.go:4]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/pkg/reg/adapter/dockerhub/consts.go [hunk 20]
```diff
@@ -1,7 +1,6 @@
 package dockerhub
 
 import "fmt"
-
```
Comment (zyyw) [inline src/pkg/reg/adapter/dockerhub/consts.go:4]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/pkg/reg/adapter/helmhub/consts.go [hunk 21]
```diff
@@ -1,7 +1,6 @@
 package helmhub
 
 import "fmt"
-
```
Comment (zyyw) [inline src/pkg/reg/adapter/helmhub/consts.go:4]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/pkg/reg/model/policy.go [hunk 22]
```diff
@@ -15,7 +15,6 @@
 package model
 
 import "github.com/goharbor/harbor/src/lib/errors"
-
```
Comment (zyyw) [inline src/pkg/reg/model/policy.go:18]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/pkg/registry/interceptor/interceptor.go [hunk 23]
```diff
@@ -15,7 +15,6 @@
 package interceptor
 
 import "net/http"
-
```
Comment (zyyw) [inline src/pkg/registry/interceptor/interceptor.go:18]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/chartserver/handler_repo_test.go (hunk 3 above)
Comment (daixiang0) [reply src/chartserver/handler_repo_test.go:4]: Now all linters do not config the one-line import, they will keep it as they were. I recommend that we keep the same with them.

The emply line is necessary, I will update them all but keep one-line import as they were.

**Comment by daixiang0** [comment]
A new release of golangci-lint will come next day, I will use the latest one to update.

File: src/pkg/retention/models.go [hunk 24]
```diff
@@ -15,7 +15,6 @@
 package retention
 
 import "time"
-
```
Comment (zyyw) [inline src/pkg/retention/models.go:18]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/pkg/scan/export/export_data_selector.go [hunk 25]
```diff
@@ -1,7 +1,6 @@
 package export
 
 import "github.com/bmatcuk/doublestar"
-
```
Comment (zyyw) [inline src/pkg/scan/export/export_data_selector.go:4]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/pkg/signature/notary/model/model.go [hunk 26]
```diff
@@ -15,7 +15,6 @@
 package model
 
 import "github.com/theupdateframework/notary/tuf/data"
-
```
Comment (zyyw) [inline src/pkg/signature/notary/model/model.go:18]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/server/middleware/immutable/error.go [hunk 27]
```diff
@@ -1,7 +1,6 @@
 package immutable
 
 import "fmt"
-
```
Comment (zyyw) [inline src/server/middleware/immutable/error.go:4]: same comment as this one:
* https://github.com/goharbor/harbor/pull/17131/files#r923010380

File: src/chartserver/cache.go (hunk 2 above)
Comment (zyyw) [reply src/chartserver/cache.go:12]: You are absolutely right!

File: src/chartserver/handler_repo_test.go (hunk 3 above)
Comment (zyyw) [reply src/chartserver/handler_repo_test.go:4]: Cool, thanks a lot!

**Comment by zyyw** [comment]
> A new release of golangci-lint will come next day, I will use the latest one to update.

Thank you! When you update the code, could you please rebase the latest main into your branch to resolve the conflicts as well?

**Comment by daixiang0** [comment]
@zyyw I have rebased the latest main.

Thanks for your hard review, the changes are big while meaningful, after this, we can get clear and ordered imports.

File: src/common/http/transport.go [hunk 28]
```diff
@@ -16,12 +16,15 @@ package http
 
 import (
 	"crypto/tls"
-	"github.com/goharbor/harbor/src/lib/trace"
 	"net"
 	"net/http"
 	"time"
 
+	"github.com/goharbor/harbor/src/lib/trace"
```
Comment (zyyw) [inline src/common/http/transport.go:23]: line 23 & line 27 are the same.

File: src/pkg/registry/interceptor/readonly/interceptor.go [hunk 29]
```diff
@@ -17,11 +17,13 @@ package readonly
 import (
 	"context"
 	"errors"
-	"github.com/goharbor/harbor/src/lib/cache/memory"
 	"net/http"
 	"time"
 
+	"github.com/goharbor/harbor/src/lib/cache/memory"
+
```
Comment (zyyw) [inline src/pkg/registry/interceptor/readonly/interceptor.go:24]: line 23 and line 26 are the same

File: src/controller/event/metadata/quota.go [hunk 30]
```diff
@@ -1,12 +1,14 @@
 package metadata
 
 import (
-	proModels "github.com/goharbor/harbor/src/pkg/project/models"
 	"time"
 
+	proModels "github.com/goharbor/harbor/src/pkg/project/models"
```
Comment (zyyw) [inline src/controller/event/metadata/quota.go:6]: line 6 and line 11 are the same

**Comment by daixiang0** [comment]
Goimports do not remove duplicate imports, let me fix it by hand :(

**Comment by zyyw** [comment]
Hi @daixiang0, I'm sorry, maybe we should stay on golangci-lint@v1.45.2, since golangci-lint@v1.47.0 does not works well. Here is my reasons:
* some files have duplicate import.
* there are 890+ files updated in yesterday's PR review while there are only 315 files updated in today's PR review. I suspect there are some files are not grouped into 3 parts (which is our primary goal of this PR, sorting import within each group is secondary and doing it in a separate PR is also acceptable)
* concerns about introducing failures of the existing enabled lint check, https://github.com/goharbor/harbor/runs/7402038562?check_suite_focus=true

**Comment by daixiang0** [comment]
OK, I will turn back to 1.45.2.

**Comment by daixiang0** [comment]
The root cause of CI failure is `goharbor/mockery` would generate some codes with unordered imports, I am not familiar with it, I will ignore those files, what do you think about it?

**Comment by daixiang0** [comment]
Has fixed, `goimports` would add new line before a comment
```
diff --git a/src/controller/icon/controller.go b/src/controller/icon/controller.go
index 9122acdab..8927aa136 100644
--- a/src/controller/icon/controller.go
+++ b/src/controller/icon/controller.go
@@ -19,6 +19,7 @@ import (
        "context"
        "encoding/base64"
        "image"
+
        // import the gif format
        _ "image/gif"
        // import the jpeg format
diff --git a/src/pkg/reg/manager.go b/src/pkg/reg/manager.go
index a316641c2..af1f57c4c 100644
--- a/src/pkg/reg/manager.go
+++ b/src/pkg/reg/manager.go
@@ -22,6 +22,7 @@ import (
        "github.com/goharbor/harbor/src/lib/config"
        "github.com/goharbor/harbor/src/lib/q"
        "github.com/goharbor/harbor/src/pkg/reg/adapter"
+
        // register the AliACR adapter
        _ "github.com/goharbor/harbor/src/pkg/reg/adapter/aliacr"
        // register the Artifact Hub adapter

```

**Comment by zyyw** [comment]
Thank you for the fix. 
Regarding the `goharbor/mockery` thing, it seems that my initial idea doesn't work. Let's ignore those mock files.

We can ignore all files under `src/testing` directory. And in terms of the following mock files that are not under `src/testing` directory, let's use pattern matching (if possible), for example: `src/controller/replication/flow/mock_*.go`,  `src/controller/replication/mock_*.go`(just an example of the idea, the matching pattern may not exactly as they are) :
```
src/controller/replication/flow/mock_adapter_factory_test.go
src/controller/replication/flow/mock_adapter_test.go
src/controller/replication/mock_flow_controller_test.go
src/jobservice/mgt/mock_manager.go
src/pkg/scheduler/mock_dao_test.go
src/pkg/task/mock_execution_dao_test.go
src/pkg/task/mock_jobservice_client_test.go
src/pkg/task/mock_task_dao_test.go
src/pkg/task/mock_task_manager_test.go
```

**Review by zyyw — APPROVED** [review]
lgtm

**Review by wy65701436 — APPROVED** [review]
lgtm

**Pull request merged**